**Notice**

- 아래와 같이 **주석으로 구분된 곳에서 코드를 수정 및 작성** 해주세요

```python
    # ================================================================== #
    #         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
    # ================================================================== #
    #
    #                이 곳에서 코드를 수정 및 작성하시오.
    #
    # ================================================================== #
    #          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
    # ================================================================== #
```

> 코드를 작성할 때에는 잠시 비디오를 멈추고 혼자 충분히 고민해보며 작성해본 후, 다시 강의를 들어주세요.


In [ ]:
# 아래 실습에 필요한 library들
from IPython.display import Image
import numpy as np
import pandas as pd

<img src="https://upload.wikimedia.org/wikipedia/commons/thumb/e/ed/Pandas_logo.svg/1280px-Pandas_logo.svg.png"
     width="250">

# Pandas

## 자료구조


## 개요

Pandas는 **관계형** 또는 **레이블이 된** 데이터로 쉽고 직관적 으로 작업할 수 있도록 설계되었고, 빠르고, 유연한 데이터 구조를 제공하는 Python 패키지이다.

또한, 어떤 언어로도 사용할 수 있는 가장 **강력하고 유연한 오픈 소스 데이터 분석 / 조직 도구**이다.

다음의 종류의 데이터에 **적합한 분석 패키지**이다.

> SQL 테이블 또는 Excel 스프레드 시트에서와 같은 열과 행으로 이루어진 테이블 형식 데이터
>
> 정렬되고 정렬되지 않은 시계열 데이터
>
> 다른 형태의 관찰 / 통계 데이터 세트

본 강의에서는 Pandas 자료구조를 정의부터 시작해서 DataFrame 활용까지 다루고자 한다.

## Pandas 공식 문서

공식 문서는 다음 링크에서 확인할 수 있다.
- [공식 도큐먼트](https://pandas.pydata.org/docs/reference/index.html)

In [ ]:
import pandas as pd
pd
print('pandas version : ', pd.__version__)

pandas version :  2.2.2


우선 Pandas를 활용하기 위해 데이터를 생성하는 방법을 알아보자.

## 1. Series

Pandas의 Series는 1차원 배열로서 다음의 특징을 가진다.

자세한 내용은 다음 링크에서 확인할 수 있다.

[도큐먼트](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.Series.html?highlight=series#pandas.Series)

> 데이터를 담는 **차원 배열 구조**를 가진다.
>
> **인덱스(index)를 사용 가능**하다.
>
> **데이터 타입**을 가진다. (dtype)


numpy로 1차원 배열 생성 후 Series로 변환하거나, list로 생성이 가능하다.

 ( numpy의 자세한 내용은 다음 강의에서 다루니, 이해하지 않아도 된다. )

In [ ]:
arr = np.arange(100, 105)
arr

array([100, 101, 102, 103, 104])

In [ ]:
s = pd.Series(arr)
s

0    100
1    101
2    102
3    103
4    104
dtype: int64

`dtype`을 지정하여 생성도 가능하다.

In [ ]:
s = pd.Series(arr, dtype='int32')
s

0    100
1    101
2    102
3    103
4    104
dtype: int32

list로도 생성이 가능하다.

In [ ]:
s = pd.Series(['부장', '차장', '대리', '사원', '인턴'])
s

0    부장
1    차장
2    대리
3    사원
4    인턴
dtype: object

Series에 다양한 데이터 타입의 데이터로도 생성이 가능하다.

생성시, **object** 타입으로 생성된다.

In [ ]:
s = pd.Series([91, 2.5, '스포츠', 4, 5.16])
s

0      91
1     2.5
2     스포츠
3       4
4    5.16
dtype: object

이제 다음의 출력 결과가 나오도록 직접 Series로 생성해보자.

s <p><strong>[출력 결과]</strong></p>
<pre>0     3.0
1     5.0
2     7.0
3     9.0
4    11.0
dtype: float32</pre>

p <p><strong>[출력 결과]</strong></p>
<pre>0    가
1    나
2    다
3    라
4    마
dtype: object</pre>

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #



s = ...
p = ...

# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #

s
p

Ellipsis

다음은 생성된 Series를 활용할 수 있는 방법을 알아보자.

## 2. 인덱싱 (indexing)

Series 생성시 0부터 순차적으로 부여되는 index를 확인할 수 있다.

이를 `RangeIndex`라 부른다.

> Start: 첫 번째 인덱스 값
>
> Stop: 데이터의 길이 (len(data))
>
> Step: 인덱스 간의 증가 값 (기본값 1)

이때, index의 값으로 Series의 특정 위치의 요소를 불러올 수 있다.

In [ ]:
s = pd.Series(['부장', '차장', '대리', '사원', '인턴'])
s
s.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
s[2]

'대리'

### fancy indexing

**팬시 인덱싱(Fancy Indexing)**은 정수 리스트나 배열을 사용하여 여러 개의 특정 행이나 열을 한 번에 선택하는 방법이다.

일반적인 단일 인덱싱과 달리, 리스트나 배열을 사용하여 특정 위치의 요소를 선택할 수 있다.

In [ ]:
s[[1, 3]]

1    차장
3    사원
dtype: object

In [ ]:
s[np.arange(1, 4, 2)]

1    차장
3    사원
dtype: object

### boolean indexing

조건식을 만들어서 특정 조건에 대하여 `True`에 해당하는 값만 필터링 할 수 있다.

s를 정의 후

In [ ]:
np.random.seed(0)
s = pd.Series(np.random.randint(10000, 20000, size=(10,)))
s

0    12732
1    19845
2    13264
3    14859
4    19225
5    17891
6    14373
7    15874
8    16744
9    13468
dtype: int64

다음과 같이 True, False로 이루어진 Boolean series를 생성 후 index로 활용하여 필터할 수 있다.

필터는 다음과 같이 사용하면 True인 값을 가져올 수 있다.

In [ ]:
s > 15000

0    False
1     True
2    False
3    False
4     True
5     True
6    False
7     True
8     True
9    False
dtype: bool

In [ ]:
s[s > 15000]

1    19845
4    19225
5    17891
7    15874
8    16744
dtype: int64

기본 값으로 부여되는 `RangeIndex`에 사용자 정의의 `index`를 **지정**할 수 있다.

In [ ]:
s = pd.Series(['마케팅', '경영', '개발', '기획', '인사'], index=['a', 'b', 'c', 'd', 'e'])
s

a    마케팅
b     경영
c     개발
d     기획
e     인사
dtype: object

In [ ]:
s.index

Index(['a', 'b', 'c', 'd', 'e'], dtype='object')

사용자 정의의 index 부여시 **변경된 index**로 조회 가능하다.

In [ ]:
s['c']

'개발'

In [ ]:
s[['a', 'd']]

a    마케팅
d     기획
dtype: object

정의할 때 지정하는 방법 외에,
먼저 `Series`를 생성 후 index 속성 값에 새로운 index를 할당하여 인덱스를 지정할 수도 있다.

In [ ]:
s = pd.Series(['마케팅', '경영', '개발', '기획', '인사'])
s.index

RangeIndex(start=0, stop=5, step=1)

In [ ]:
s.index = list('abcde')
s

a    마케팅
b     경영
c     개발
d     기획
e     인사
dtype: object

이제, 직접 원하는 데이터를 필터링해보자.

다음과 같이 정의된
`sample1` 중 160 이하인 데이터만 남겨보자

In [ ]:
np.random.seed(20)
sample1 = pd.Series(np.random.randint(100, 200, size=(15,)))
sample1

0     199
1     190
2     115
3     195
4     128
5     190
6     109
7     120
8     175
9     122
10    171
11    134
12    196
13    140
14    185
dtype: int64

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #



# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #

## 3. 속성 (attribute)

### values

`values`는 Series 데이터 값(value)만 **numpy array** 형식으로 가져 온다.

In [ ]:
s.values

array(['마케팅', '경영', '개발', '기획', '인사'], dtype=object)

### ndim - 차원
Series는 1차원 자료구조이기 때문에 ndim 출력시 **1**이 출력된다.

In [ ]:
s.ndim

1

### shape

shape은 데이터의 모양(shape)을 알아보기 위하여 사용하는데, Series의 shape은 **데이터의 갯수**를 나타낸다.

**튜플(tuple)** 형식으로 출력된다.

In [ ]:
s.shape

(5,)

### NaN (Not a Number)

Pandas에서 **NaN 값**은 비어있는 **결측치 데이터**를 의미합니다.

임의로 비어있는 값을 대입하고자 할 때는 **numpy의 nan (np.nan)**을 입력합니다.

In [ ]:
s = pd.Series(['선화', '강호', np.nan, '소정', '우영'])
s

0     선화
1     강호
2    NaN
3     소정
4     우영
dtype: object

### 결측치 (NaN) 값 처리

`isnull()`과 `isna()`은 **NaN** 값을 찾는 함수이다.

`isnull()`과 `isna()`는 결과가 동일하다.

In [ ]:
s.isnull()

0    False
1    False
2     True
3    False
4    False
dtype: bool

In [ ]:
s.isna()

0    False
1    False
2     True
3    False
4    False
dtype: bool

이를 boolean indexing에 적용해볼 수 있다.

In [ ]:
s[s.isnull()]

2    NaN
dtype: object

In [ ]:
s[s.isna()]

2    NaN
dtype: object

`notnull()`은 NaN값이 아닌, 즉 비어있지 않은 데이터를 찾는 함수 입니다.

In [ ]:
s.notnull()

0     True
1     True
2    False
3     True
4     True
dtype: bool

In [ ]:
s.notna()

0     True
1     True
2    False
3     True
4     True
dtype: bool

NaN 값을 제외한, 나머지 값을 가져오려면 다음과 같이 할 수 있다.

In [ ]:
s[s.notnull()]

0    선화
1    강호
3    소정
4    우영
dtype: object

이제, `sample` 중 결측치 데이터만 필터해보자

In [ ]:
sample = pd.Series(['IT서비스', np.nan, '반도체', np.nan, '바이오', '자율주행'])
sample

0    IT서비스
1      NaN
2      반도체
3      NaN
4      바이오
5     자율주행
dtype: object

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #



# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #


## 4. 슬라이싱

원하는 index의 데이터를 연속으로 가져올 수 있다.

**(주의)** 숫자형 index로 접근할 때는 뒷 index가 포함되지 않는다.

In [ ]:
s = pd.Series(np.arange(100, 150, 10))
s

0    100
1    110
2    120
3    130
4    140
dtype: int64

In [ ]:
s[1:3]

1    110
2    120
dtype: int64

새롭게 지정한 인덱스(문자열)는 시작 index와 끝 index **모두 포함**한다.

In [ ]:
s.index = list('가나다라마')
s

가    100
나    110
다    120
라    130
마    140
dtype: int64

In [ ]:
s['나':'라']

나    110
다    120
라    130
dtype: int64

`sample`에서 다음과 같은 결과를 가지도록 **슬라이싱** 해보자.

<p><strong>[출력 결과]</strong></p>
<pre>2    164
3    167
4    167
5    109
6    183
dtype: int64</pre>

In [ ]:
np.random.seed(0)
sample = pd.Series(np.random.randint(100, 200, size=(10,)))
sample

0    144
1    147
2    164
3    167
4    167
5    109
6    183
7    121
8    136
9    187
dtype: int64

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #


# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #

`sample2`에서 다음과 같은 결과를 가지도록 **슬라이싱** 해보자

<p><strong>[출력 결과]</strong></p>
<pre>바    109
사    183
아    121
자    136
차    187
dtype: int64</pre>

In [ ]:
np.random.seed(0)
sample2 = pd.Series(np.random.randint(100, 200, size=(10,)), index=list('가나다라마바사아자차'))
sample2

가    144
나    147
다    164
라    167
마    167
바    109
사    183
아    121
자    136
차    187
dtype: int64

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #


# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #


## 5. DataFrame

[도큐먼트](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.html)

1차원 데이터 구조 보다는, 주로 많이쓰이는 Excel 데이터 시트와 같은 구조를 가지는 DataFrame에 대해서 알아보자.

`pd.DataFrame`

> 2차원 데이터 구조 (Excel 데이터 시트를 생각하면 된다)
>
> 행(row), 열(column)으로 구성되어 있다.
>
> 각 열(column)은 각각의 데이터 타입 (dtype)을 가진다.

### 생성

**list 를 통해 생성**할 수 있다. DataFrame을 만들 때는 **2차원 list를 대입**한다.

In [ ]:
pd.DataFrame([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]])

0  1  2
0  1  2  3
1  4  5  6
2  7  8  9

아래 예제와 같이 **columns를 지정**하면, DataFrame의 각 열에 대한 컬럼명이 붙는다.

In [ ]:
pd.DataFrame([[1, 2, 3],
              [4, 5, 6],
              [7, 8, 9]], columns=['가', '나', '다'])

가  나  다
0  1  2  3
1  4  5  6
2  7  8  9

**dictionary를 통한 생성**도 가능하다.

편리한 점은 dictionary의 **key 값이 자동으로 column 명으로 지정**된다는 점이다.

In [ ]:
data = {
    'name': ['Kim', 'Lee', 'Park'],
    'age': [24, 27, 34],
    'children': [2, 1, 3]
}
df = pd.DataFrame(data)
df

name  age  children
0   Kim   24         2
1   Lee   27         1
2  Park   34         3

### 속성 (Attribute)

DataFrame은 다음의 **속성**을 가진다.

- **index**: index (기본 값으로 RangeIndex)
- **columns**: column 명
- **values**: numpy array형식의 데이터 값
- **dtypes**: column 별 데이터 타입
- **T**: DataFrame을 전치(Transpose)

In [ ]:
df.index

RangeIndex(start=0, stop=3, step=1)

In [ ]:
df.columns

Index(['name', 'age', 'children'], dtype='object')

In [ ]:
df.values

array([['Kim', 24, 2],
       ['Lee', 27, 1],
       ['Park', 34, 3]], dtype=object)

데이터프레임에서의 `dtypes` 속성은 컬럼별 dtype을 출력한다.

In [ ]:
df.dtypes

name        object
age          int64
children     int64
dtype: object

In [ ]:
df.T

0    1     2
name      Kim  Lee  Park
age        24   27    34
children    2    1     3

### index 지정

데이터프레임에서도 Series와 마찬가지로 index를 지정할 수 있다.

In [ ]:
df

name  age  children
0   Kim   24         2
1   Lee   27         1
2  Park   34         3

In [ ]:
df.index = list('abc')
df

name  age  children
a   Kim   24         2
b   Lee   27         1
c  Park   34         3

(참고) DataFrame의 indexing / slicing은 나중에 세부적으로 다루도록 하겠다.

### column 다루기

DataFrame에 key 값으로 column의 이름을 지정하여 column을 선택할 수 있다.

1개의 column을 가져올 수 있으며, **1개의 column 선택시 Series**가 된다.

In [ ]:
df['name']

a     Kim
b     Lee
c    Park
Name: name, dtype: object

In [ ]:
type(df['name'])

pandas.core.series.Series

2개 이상의 column 선택은 **fancy indexing으로 가능**하다.

In [ ]:
df[['name', 'children']]

name  children
a   Kim         2
b   Lee         1
c  Park         3

(참고) column에 대한 slicing도 가능 하지만 이 부분도 나중에 다루도록 하겠다.

**rename**으로 column명 변경 가능하다.


DataFrame.rename(columns={'바꾸고자 하는 컬럼명': '바꿀 컬럼명'})

In [ ]:
df.rename(columns={'name': '이름'})

이름  age  children
a   Kim   24         2
b   Lee   27         1
c  Park   34         3

`inplace=True` 옵션으로 변경사항을 바로 적용할 수 있다.

In [ ]:
df.rename(columns={'name': '이름'}, inplace=True)
df

이름  age  children
a   Kim   24         2
b   Lee   27         1
c  Park   34         3

이제 직접 DataFrame을 생성해보자.

다음 출력결과가 나오도록 df를 생성해보자.

- 생성된 DataFrame은 df 변수에 할당한다.

<p><strong>[출력 결과]</strong></p>
<div>
<style scoped>
    .dataframe tbody tr th:only-of-type {
        vertical-align: middle;
    }

    .dataframe tbody tr th {
        vertical-align: top;
    }

    .dataframe thead th {
        text-align: right;
    }
</style>
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>food</th>
      <th>price</th>
      <th>rating</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>0</th>
      <td>KFC</td>
      <td>1000</td>
      <td>4.5</td>
    </tr>
    <tr>
      <th>1</th>
      <td>McDonald</td>
      <td>2000</td>
      <td>3.9</td>
    </tr>
    <tr>
      <th>2</th>
      <td>SchoolFood</td>
      <td>2500</td>
      <td>4.2</td>
    </tr>
  </tbody>
</table>
</div>

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #



# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #

df

food 컬럼과 rating 컬럼만 선택하여 출력해보자.

In [ ]:
# ================================================================== #
#         START OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)        #
# ================================================================== #



# ================================================================== #
#          END OF YOUR CODE (DO NOT DELETE/MODIFY THIS LINE)         #
# ================================================================== #
